CODING_RULES.md

```diff
@@ -11,90 +11,156 @@ to be completed
 
 - use getters/setters instead of `@property`
 
-### regarding the exceptions
-
-
-- use exceptions defined in **fedbiomed.common.exceptions**
-
-- callee side: then detecting a python exception in a fedbiomed layer :
-
-  - print a logger.*() message
-
-    - could be logger.critical -> stop the software, cannot continue
-      (ex: FedbiomedEnvironError)
-    - or logger.error -> software may continue (ex: MessageError)
-
-  - raise the exception as a fedbiomed exception
-
-
-- caller/intermediate side: trap the exceptions:
-
-  - separate the FedbiomedError and other exceptions
-
-```
-try:
-
-    something()
-
-except SysError as e:
-    logger.XXX()
-    raise OneOfFedbiomedError()
-```
-
-  - example of top level function (eg: Experiment())
-
-```
-try:
-
-    something()
-
-except FedbiomedError as e:
-    etc...
-
-except Exception as e:   <=== objective: minimize the number of type we arrive here !
-    # place to do a backtrace
-    # extra message to the end user to post this backtrace to the support team
-    etc...
-```
-
-  - except of the top level program, it is **forbidden** to trap all exceptions (with ```except:``` or ```except Exception```)
-
-
-- the **try:** block is as small as possible
-
-- force to read the documentation
-
+### exceptions handling
+
+- on the node: in general, node should not stop because of exceptions that occur while executing requests received from researcher
+
+- on the researcher: general behaviour is to propagate the exceptions to the top level layer, where they are transformed to a friendlier output. Researcher displays this output and stops.
+
+- when a class raises an exception, it raises a FedbiomedSomethingError: use exceptions defined in **fedbiomed.common.exceptions** :
+
+  Do:
+  ```
+  raise FedbiomedSometypeError()
+  ```
+
+  Don't:
+  ```
+  raise NameError()
+  ```
+
+- optionally, if more specificity is wanted, a class can catch a python (non-Fed-Biomed) exception and re-raise it as a FedbiomedError
+
+  Optionally do:
+  ```
+  try:
+      something()
+  except SysError as e:
+      raise OneOfFedbiomedError()
+  ```
+
+- a class generally shouldn't catch a FedbiomedError and re-raise (a FedbiomedError):
+
+  Don't:
+  ```
+  try:
+    somecode()
+  except FedbiomedSometypeError;
+    raise FedbiomedOnetypeError()
+  ```
+
+- when catching exceptions
+
+  - try to be specific about the exception if easy/meaningful:
+
+    Ideally:
+    ```
+    try:
+      mycode()
+    except ValueError as e:
+      ...
+    ```
+
+    If needed:
+    ```
+    try:
+      mycode()
+    except Exception as e:
+      ...
+    ```
+
+  - don't use the `except:` clause
+
+    Don't
+    ```
+    try:
+      mycode()
+    except:
+      ...
+    ```
+
+  - should separate FedbiomedError and other exceptions
+
```

Comment:
ACtually it depends what the dev wants to do: he could just want to log a generic error or process wrt the error raised. I would complete the example:

```
    try:
      mycode()
    except FedbiomedSomeError as e:
       logger.warning("mycode failed with SomeError")
    except FedbiomedError as e:
       logger.critical("mycode failed with a FedBioMedError")
    except Exception as e:
       logger.critical(f"mycode failed with unknown error {e}")

```

Reply:
Thank you @ybouilla 

OK to add a mention like "so as to take distinct actions".
But I don't like the idea of writing `logger.xxx()` in the example because it can confuse the dev into thinking that he should always log message in such scenario (while it can also raise error, do none of the above). In other words, this rule is not about logging IMO.